File: backend-nestjs/src/llm/adapters/anthropic.adapter.ts
import { Injectable } from '@nestjs/common';
import { GenerateInput, GenerateOptions, GenerateResult, LlmAdapter } from './llm.adapter';

@Injectable()
export class AnthropicAdapter implements LlmAdapter {
  readonly provider = 'claude' as const;

  async validateKey(apiKey: string): Promise<boolean> {
    return apiKey.startsWith('sk-') && apiKey.length > 10;
  }

  async generateThread(apiKey: string, input: GenerateInput, options: GenerateOptions): Promise<GenerateResult> {
    void apiKey;
    return {
      providerUsed: this.provider,
      model: options.model ?? 'sonnet',
      hook: `${input.title} 핵심 정리`,
      points: ['핵심 포인트 1', '핵심 포인트 2', '핵심 포인트 3'],
      insight: '후속 시장 반응 추적이 필요합니다.',
      hashtags: ['#이슈정리', '#트렌드', '#뉴스'],
      source: input.url,
      tokenIn: 1200,
      tokenOut: 280,
    };
  }
}

File: backend-fastapi/app/adapters/anthropic_adapter.py
from __future__ import annotations

import httpx

from app.adapters.base import (
    GenerateInput, GenerateOptions, GenerateResult, LLMAdapter,
    build_thread_prompt, parse_thread_json,
)

_ANTHROPIC_API_URL = 'https://api.anthropic.com/v1/messages'
_DEFAULT_MODEL = 'claude-sonnet-4-20250514'


class AnthropicAdapter(LLMAdapter):
    provider_name = 'claude'

    async def validate_key(self, api_key: str) -> bool:
        return api_key.startswith('sk-') and len(api_key) > 10

    async def generate_thread(self, api_key: str, data: GenerateInput, opt: GenerateOptions) -> GenerateResult:
        model = opt.model or _DEFAULT_MODEL
        prompt = build_thread_prompt(data, opt)

        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                _ANTHROPIC_API_URL,
                headers={
                    'x-api-key': api_key,
                    'anthropic-version': '2023-06-01',
                    'content-type': 'application/json',
                },
                json={
                    'model': model,
                    'max_tokens': 1024,
                    'messages': [{'role': 'user', 'content': prompt}],
                },
            )

        if resp.status_code != 200:
            detail = resp.text[:300]
            raise RuntimeError(f'Anthropic API error {resp.status_code}: {detail}')

        body = resp.json()
        raw_text = body.get('content', [{}])[0].get('text', '')
        usage = body.get('usage', {})
        parsed = parse_thread_json(raw_text)

        return GenerateResult(
            provider_used=self.provider_name,
            model=model,
            hook=parsed.get('hook', ''),
            points=parsed.get('points', []),
            insight=parsed.get('insight', ''),
            hashtags=parsed.get('hashtags', []),
            source=data.url,
            token_in=usage.get('input_tokens', 0),
            token_out=usage.get('output_tokens', 0),
        )
